contracts/serverless-v2/CommonChainJobs.sol

```diff
@@ -0,0 +1,357 @@
+        // reward ratio - 2:1:0
+    }
+
+    function isJobExecutor(
+        uint256 _jobId,
+        uint256 _reqChainId,
+        address _executor
+    ) public view returns (bool) {
+        uint256 key = getKey(_jobId, _reqChainId);
+        address[] memory selectedNodes = selectedExecutors[key];
+        uint256 len = selectedExecutors[key].length;
+        for (uint256 index = 0; index < len; index++) {
+            if(selectedNodes[index] == _executor)
+                return true;
+        }
+        return false;
+    }
+
+    //-------------------------------- Job end --------------------------------//
+
+    //-------------------------------- Timeout start --------------------------------//
+
+    event SlashedOnExecutionTimeout(
+        uint256 indexed jobId,
+        uint256 indexed reqChainId,
+        address[] executors
+    );
+
+    event GatewayReassigned(
+        uint256 indexed jobId,
+        uint256 indexed reqChainId,
+        address prevGatewayOperator,
+        address newGatewayOperator
+    );
+
+    function slashOnExecutionTimeout(
+        uint256 _jobId,
+        uint256 _reqChainId
+    ) external {
+        uint256 key = getKey(_jobId, _reqChainId);
+        require(jobs[key].jobId > 0, "INVALID_JOB");
+
+        // check for time
+        require(
+            block.timestamp > jobs[key].execStartTime + jobs[key].deadline + executionBufferTime,
+            "DEADLINE_NOT_OVER"
+        );
+
+        delete jobs[key];
+
+        // slash Execution node
+
+        uint256 len = selectedExecutors[key].length;
+        for (uint256 index = 0; index < len; index++) {
+            address executorKey = selectedExecutors[key][index];
+            require(!hasExecutedJob[key][executorKey], "JOB_ALREADY_EXECUTED");
+            delete hasExecutedJob[key][executorKey];
+            executors.updateOnExecutionTimeoutSlash(executorKey);
+        }
+
+        emit SlashedOnExecutionTimeout(_jobId, _reqChainId, selectedExecutors[key]);
+
+        delete selectedExecutors[key];
+    }
+
+    function reassignGatewayRelay(
+        address _gatewayOperatorOld,
+        uint256 _jobId,
+        uint256 _reqChainId,
+        bytes memory _signature,
+        uint8 _sequenceId
+    ) external {
+        uint256 key = getKey(_jobId, _reqChainId);
+        // time check will be done in the gateway enclaves and based on the algo, a new gateway will be selected
+
+        require(!jobs[key].isResourceUnavailable, "JOB_MARKED_ENDED_AS_RESOURCE_UNAVAILABLE");
+        require(_sequenceId == jobs[key].sequenceId + 1 && _sequenceId < 3, "INVALID_SEQUENCE_ID");
+        jobs[key].sequenceId = _sequenceId;
+
+        // signature check
+        bytes32 digest = keccak256(abi.encode(_jobId, _reqChainId, _gatewayOperatorOld, _sequenceId));
+        address signer = digest.recover(_signature);
+
+        gateways.allowOnlyVerified(signer);
+
+        emit GatewayReassigned(_jobId, _reqChainId, _gatewayOperatorOld, _msgSender());
```

Comment:
an you add `_sequenceId` to the event to know which retry it is. Will help in clearing the `active_jobs` in the enclave. Otherwise, will cause issues in the case when the same gateway again after slashing.